x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformPersistentTasksExecutor.java

```diff
@@ -84,7 +85,18 @@ public DataFrameTransformPersistentTasksExecutor(Client client,
             logger.debug(reason);
             return new PersistentTasksCustomMetaData.Assignment(null, reason);
         }
-        return super.getAssignment(params, clusterState);
+        if (clusterState.nodes().getMinNodeVersion().onOrAfter(params.getVersion())) {
+            return super.getAssignment(params, clusterState);
+        } else {
+            DiscoveryNode discoveryNode = selectLeastLoadedNode(clusterState, (node) ->
+                node.isDataNode() && node.getVersion().onOrAfter(params.getVersion())
+            );
+            if (discoveryNode == null) {
+                return NO_NODE_FOUND;
+            } else {
+                return new PersistentTasksCustomMetaData.Assignment(discoveryNode.getId(), "");
+            }
+        }
```

Comment:
The code in the `else` block here is basically the same as `super.getAssignment` but with the extra requirement of a node version.  However, if someone altered the super class code without realising this override existed then the logic of the `else` block would deviate from the super class, and that could lead to weird assignments during rolling upgrades.  So I think it would be better to scrap the `getMinNodeVersion` check and _always_ use the code that's in the `else` block.